### PR TITLE
Stricter type definitions for dex data

### PR DESCRIPTION
**Pull request by Slayer95**
- All event handlers typed
- BattleStatuses switched to PureEffectData

### CODE REVIEW

**Comment by Zarel** [comment]
This pull request **introduces 4 alerts** when merging 9d5a75c381bc190cb0d10d4b8832fdc2f0d87c49 into f786b14b4c39b5336befae5170c4b1beff234e1e - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-eaf6155d68bd5ff6648e045a1e8fed3130b698b7)

**new alerts:**

* 4 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Slayer95** [comment]
Damn, so fussy ! Will fix when I wake up

File: data/moves.js [hunk 1]
```diff
@@ -3249,9 +3246,9 @@ let BattleMovedex = {
 		priority: 0,
 		flags: {protect: 1, reflectable: 1, mirror: 1, authentic: 1},
 		onHit(target, source, move) {
-			/**@type {?boolean | number} */
+			/**@type {boolean} */
```
Comment (scheibo) [inline data/moves.js]: this should be inferrable?

File: dev-tools/globals.ts [hunk 2]
```diff
@@ -155,176 +155,490 @@ interface SecondaryEffect {
...
+	onTakeItem?: ((this: Battle, item: Item, pokemon: Pokemon, source: Pokemon, move?: ActiveMove) => boolean | void) | boolean
+}
+
+interface MoveEventMethods {
 	/** Return true to stop the move from being used */
 	beforeMoveCallback?: (this: Battle, pokemon: Pokemon, target: Pokemon | null, move: ActiveMove) => boolean | void
 	beforeTurnCallback?: (this: Battle, pokemon: Pokemon, target: Pokemon) => void
 	damageCallback?: (this: Battle, pokemon: Pokemon, target: Pokemon) => number | false
+
+	onAfterHit?: CommonHandlers["VoidSourceMove"]
```
Comment (scheibo) [inline dev-tools/globals.ts]: does Typescript not like `'`? I dont know by our string rules that `"` is approriate?

Comment (Slayer95) [reply dev-tools/globals.ts]: This file was already using double quotes, but originally it should use single quotes, yes.

Comment (Slayer95) [reply dev-tools/globals.ts]: Or, better-yet, dot-notation

File: data/mods/gen1/moves.js [hunk 3]
```diff
@@ -204,10 +201,7 @@ let BattleMovedex = {
 		self: {
 			volatileStatus: 'partialtrappinglock',
 		},
-		onBeforeMove(pokemon, target, move) {
-			// Removes must recharge volatile even if it misses
-			target.removeVolatile('mustrecharge');
-		},
+		// FIXME: onBeforeMove(pokemon, target) {target.removeVolatile('mustrecharge')}
```
Comment (Zarel) [inline data/mods/gen1/moves.js]: What's going on here?

File: data/mods/gennext/moves.js [hunk 4]
```diff
@@ -296,7 +296,6 @@ let BattleMovedex = {
 		},
 		willCrit: true,
 		accuracy: true,
-		onTryHitPriority: 10,
```
Comment (Zarel) [inline data/mods/gennext/moves.js]: You've removed a lot of `Priority` entries. Can you tell me why? Are they all confirmed unnecessary?

Comment (Slayer95) [reply data/mods/gennext/moves.js]: ``moves.js`` doesn't support Priority!

File: data/mods/gen1/moves.js (hunk 3 above)
Comment (Slayer95) [reply data/mods/gen1/moves.js]: There is no ``singleEvent('BeforeMove', effect, effectData, move)``

Comment (Zarel) [reply data/mods/gen1/moves.js]: Oh, so this always did nothing? Huh.

**Pull request merged**